src/main/java/io/pravega/connectors/hadoop/AbstractPravegaBuilder.java

```diff
@@ -0,0 +1,116 @@
+/**
+ * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package io.pravega.connectors.hadoop;
+
+import org.apache.hadoop.conf.Configuration;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * An abstract pravega hadoop connector builder.
+ *
+ * The builder is abstracted to act as the base for both the {@link PravegaInputFormatBuilder} builder.
+ *
+ * @param <B> the builder type.
+ */
+abstract class PravegaBuilder<B extends PravegaBuilder> {
+
+    protected Map<String, String> settings = new HashMap<>();
+
+    private Configuration conf;
+
+    public PravegaBuilder() {}
+
+    public PravegaBuilder(Configuration conf) {
+        this.conf = conf;
+    }
+
+    /**
+     * build Configuration from key value pairs.
+     *
+     * @return Configuration associated with pravega hadoop connectors' settings.
+     */
+    public Configuration build() {
+        check();
+        if (this.conf == null) {
+            this.conf = new Configuration();
+        }
+        for (Map.Entry<String, String> entry : this.settings.entrySet()) {
+            this.conf.setStrings(entry.getKey(), entry.getValue());
+        }
+        return this.conf;
+    }
+
+    /**
+     * register key value pair
+     *
+     * @param param parameter name from {@link PravegaConfig}.
+     * @param value
+     */
+
+    protected void setString(String param, String value) {
+        this.settings.put(param, value);
+    }
+
+    /**
+     * Adds scope name to bulider
+     *
+     * @param scopeName String
+     * @return builder instance
+     */
+    public abstract B withScope(String scopeName);
+
+    /**
+     * Adds stream name to bulider
+     *
+     * @param streamName String
+     * @return builder instance
+     */
+    public abstract B forStream(String streamName);
+
+    /**
+     * Adds URI to bulider
+     *
+     * @param uri String
+     * @return builder instance
+     */
+    public abstract B withURI(String uri);
+
+    /**
+     * Adds Deserializer class name to bulider
+     *
+     * @param className String
+     * @return builder instance
+     */
+    public abstract B withDeserializer(String className);
+
+    /**
+     * Adds optional start positions to bulider
+     *
+     * @param startPos String
```

Comment:
Could you elaborate on what this start position is? Is this a stream cut in string format?

Reply:
Also, I'm thinking that if the argument is a stream cut, then it should be "position" singular given that we are passing one stream position given by the cut. I'd use plural if we are referring to multiple offsets for the segments represented in the cut.

Reply:
the "positions" isn't a stream cut in string format, but the segment/offset pairs of all segments in a single stream cut. It can be used to construct a stream cut. that's why I think we shall use a plural.

Reply:
I'm sorry if it was not clear, but I was asking you to elaborate in the javadoc, currently it only says `String`, but it does not explain what the parameter is or where it is coming from. It is important in this javadoc that we explain how the user is supposed to work with this particular builder call (and in general I think we can improve quite a bit the javadoc descriptions to give more detail, examples, etc.).